File: typescript/src/internal/registry-index.ts
/** Small, case-insensitive catalogue lookup helpers. @internal */

import { requireString } from './argument-guards.js';

/** Keys whose non-null values are strings. */
type StringField<T extends object> = {
    [K in keyof T]-?: Exclude<T[K], null | undefined> extends string ? K : never;
}[keyof T] &
    string;

/** An immutable first-record-by-key index. */
export type KeyIndex<T> = Readonly<Record<string, T>>;

/**
 * The label for the catalogue's own side of a comparison, and for a catalogue index built
 * from bundled data. It names a failure a malformed data file would have to cause, never
 * a caller: the comparison paths reach {@link normalizeKey} only once `stringField` or
 * `Object.keys` has established a string, and the index builders read a `data/` payload.
 *
 * @internal
 */
export const CATALOGUE_KEY = 'catalogue key';

/**
 * Normalize catalogue and consumer keys by trimming and folding case.
 *
 * The catalogue lookup helpers in this module funnel consumer keys through here, so
 * this is also where those lookups catch a wrong-typed key. Other case-insensitive
 * comparisons — such as structural parsing and exact game slot identifiers — define
 * their own normalization rules at their public entry points. Invalid catalogue keys
 * fail with a message that names the public parameter and received value.
 *
 * **A nullish key is not a wrong type.** It stays a miss, which is what the whole
 * lookup family answers for a key no record carries, and what an optional field an
 * import did not carry has to keep meaning — an engineering block with no
 * `ExperimentalEffect` must compare equal to a catalogue entry that has none. The
 * strict factories (`ProceduralSystem.fromName`, `ShipLoadout.empty`) are where a
 * missing argument is loud; a search is not.
 *
 * @param value - The key to normalize.
 * @param label - How to name it in a failure, `"function: parameter"` — see
 * {@link requireString}. Callers pass the *public* parameter they received, so a lookup
 * reached through a facade still names the function the consumer called.
 * @throws {TypeError} If `value` is present and not a string.
 */
export function normalizeKey(value: string, label: string): string;
/** Preserve an absent optional key while normalizing a present one. */
export function normalizeKey(value: string | undefined, label: string): string | undefined;
export function normalizeKey(value: string | undefined, label: string): string | undefined {
    return value == null ? undefined : requireString(value, label).trim().toLowerCase();
}

/**
 * Build an immutable first-record-by-key index for a fixed built-in catalogue.
 *
 * @remarks
 * This is intentionally explicit rather than cached. A domain pays for an index only
 * when it declares one for a hot built-in lookup; caller-supplied catalogues continue
 * through the scan helpers below, so mutable arrays and records never produce stale
 * answers. Duplicate keys retain the first record, matching `Array.prototype.find`.
 */
export function createKeyIndex<T extends object>(
    catalogue: readonly T[],
    field: StringField<T>,
): KeyIndex<T> {
    const index = Object.create(null) as Record<string, T>;
    for (const record of catalogue) {
        const raw = stringField(record, field);
        if (raw === null) continue;
        const key = normalizeKey(raw, CATALOGUE_KEY);
        if (!Object.hasOwn(index, key)) index[key] = record;
    }
    return Object.freeze(index);
}

/**
 * Look up a normalized key in an immutable index.
 *
 * @param label - How to name `wanted` in a failure — see {@link normalizeKey}.
 */
export function findInKeyIndex<T>(index: KeyIndex<T>, wanted: string, label: string): T | null {
    return index[normalizeKey(wanted, label)] ?? null;
}

/**
 * Look up a raw-keyed catalogue: exact key first, then a case-insensitive scan.
 *
 * @remarks
 * This is the counterpart to {@link findInKeyIndex} for a `Record` that keeps its
 * source's own casing — a Frontier `fdname` catalogue, say — rather than one built by
 * {@link createKeyIndex}. The own-property hit is both the fast path and the tie-break:
 * a catalogue holding two keys differing only in case answers the exact spelling first,
 * and only a miss pays for the scan. Inherited keys never match, so `'toString'` is a
 * miss unless the catalogue really holds it; the scan reads own *enumerable* keys, so a
 * hidden own key is reachable by its exact spelling only.
 *
 * @param label - How to name `wanted` in a failure — see {@link normalizeKey}. The key
 * is normalized before the own-property hit is tried, so a wrong-typed key fails here
 * rather than reaching the catalogue as a property name.
 */
export function findByRawKey<T>(
    catalogue: Readonly<Record<string, T>>,
    wanted: string,
    label: string,
): T | null {
    const key = normalizeKey(wanted, label);
    if (Object.hasOwn(catalogue, wanted)) return catalogue[wanted]!;
    for (const candidate of Object.keys(catalogue)) {
        if (normalizeKey(candidate, CATALOGUE_KEY) === key) return catalogue[candidate]!;
    }
    return null;
}

/**
 * Find the first matching record by scanning the supplied catalogue.
 *
 * @param label - How to name `wanted` in a failure — see {@link normalizeKey}.
 */
export function findByKey<T extends object>(
    catalogue: readonly T[],
    field: StringField<T>,
    wanted: string,
    label: string,
): T | null {
    const key = normalizeKey(wanted, label);
    return catalogue.find((record) => matches(stringField(record, field), key)) ?? null;
}

/**
 * Find every matching record by scanning the supplied catalogue.
 *
 * @param label - How to name `wanted` in a failure — see {@link normalizeKey}.
 */
export function filterByKey<T extends object>(
    catalogue: readonly T[],
    field: StringField<T>,
    wanted: string,
    label: string,
): T[] {
    const key = normalizeKey(wanted, label);
    return catalogue.filter((record) => matches(stringField(record, field), key));
}

function stringField<T extends object>(record: T, field: StringField<T>): string | null {
    const value = record[field];
    return typeof value === 'string' ? value : null;
}

function matches(raw: string | null, key: string): boolean {
    return raw !== null && normalizeKey(raw, CATALOGUE_KEY) === key;
}

File: dotnet/src/EliteDangerousAlmanac/Internal/RegistryIndex.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace EliteDangerousAlmanac.Internal;

/// <summary>
/// Case-insensitive catalogue lookup helpers.
/// </summary>
/// <remarks>
/// <para>
/// Every catalogue key — a Frontier symbol, a display name, a chemical element symbol —
/// is compared with leading and trailing whitespace removed and case folded, because a
/// player journal reports the same symbol in a different case from the catalogue.
/// </para>
/// <para>
/// A <see langword="null"/> key is a miss, not a failure. That is what a lookup answers
/// for any key no record carries, and what an optional field an import did not carry has
/// to keep meaning.
/// </para>
/// </remarks>
internal static class RegistryIndex
{
    /// <summary>The comparer every catalogue key uses once it is trimmed.</summary>
    internal static readonly StringComparer KeyComparer = StringComparer.OrdinalIgnoreCase;

    /// <summary>Trims a key, or answers <see langword="null"/> for an absent one.</summary>
    internal static string? NormalizeKey(string? value) => value?.Trim();

    /// <summary>
    /// Builds a first-record-by-key index over a fixed built-in catalogue.
    /// </summary>
    /// <remarks>
    /// Records with no value for the key are skipped, and a duplicate key keeps the first
    /// record — the answer a linear scan gives.
    /// </remarks>
    internal static IReadOnlyDictionary<string, T> CreateKeyIndex<T>(
        IReadOnlyList<T> catalogue,
        Func<T, string?> key)
    {
        Dictionary<string, T> index = new(catalogue.Count, KeyComparer);
        foreach (T record in catalogue)
        {
            string? raw = NormalizeKey(key(record));
            if (raw is null || raw.Length == 0 || index.ContainsKey(raw)) continue;
            index.Add(raw, record);
        }

        return new ReadOnlyDictionary<string, T>(index);
    }

    /// <summary>Freezes a catalogue that a data file already keys, for case-insensitive lookups.</summary>
    /// <typeparam name="T">The record shape one value has.</typeparam>
    /// <param name="raw">The catalogue as the file states it.</param>
    /// <returns>A read-only map that answers whatever case a caller writes.</returns>
    internal static ReadOnlyDictionary<string, T> FreezeByRawKey<T>(IReadOnlyDictionary<string, T> raw)
    {
        Dictionary<string, T> index = new(raw.Count, KeyComparer);
        foreach (KeyValuePair<string, T> entry in raw) index[entry.Key] = entry.Value;
        return new ReadOnlyDictionary<string, T>(index);
    }

    /// <summary>Looks a key up in an index built by <see cref="CreateKeyIndex"/>.</summary>
    /// <returns>The record, or <see langword="null"/> for a key no record carries.</returns>
    internal static T? FindInKeyIndex<T>(IReadOnlyDictionary<string, T> index, string? wanted)
        where T : class
    {
        string? key = NormalizeKey(wanted);
        if (key is null || key.Length == 0) return null;
        return index.TryGetValue(key, out T? record) ? record : null;
    }

    /// <summary>Finds the first record whose key matches, by scanning the catalogue.</summary>
    /// <returns>The record, or <see langword="null"/> for a key no record carries.</returns>
    internal static T? FindByKey<T>(IEnumerable<T> catalogue, Func<T, string?> key, string? wanted)
        where T : class
    {
        string? normalized = NormalizeKey(wanted);
        if (normalized is null || normalized.Length == 0) return null;
        foreach (T record in catalogue)
        {
            if (Matches(key(record), normalized)) return record;
        }

        return null;
    }

    /// <summary>Finds every record whose key matches, by scanning the catalogue.</summary>
    /// <returns>A read-only list of matches, in catalogue order.</returns>
    internal static IReadOnlyList<T> FilterByKey<T>(
        IEnumerable<T> catalogue,
        Func<T, string?> key,
        string? wanted)
    {
        List<T> matches = [];
        string? normalized = NormalizeKey(wanted);
        if (normalized is null || normalized.Length == 0) return new ReadOnlyCollection<T>(matches);
        foreach (T record in catalogue)
        {
            if (Matches(key(record), normalized)) matches.Add(record);
        }

        return new ReadOnlyCollection<T>(matches);
    }

    /// <summary>Answers whether a record's raw key equals an already normalized one.</summary>
    internal static bool Matches(string? raw, string normalized)
    {
        string? candidate = NormalizeKey(raw);
        return candidate is not null && KeyComparer.Equals(candidate, normalized);
    }
}
